universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java

```diff
@@ -191,6 +192,43 @@ public boolean hasEnumerator(long enumeratorId) {
         .collect(ImmutableList.toImmutableList());
   }
 
+  /**
+   * Returns a list of the question definitions that may be used to define predicates on the block
+   * definition with the given ID.
+   *
+   * <p>The available question definitions for predicates satisfy BOTH of the following:
+   *
+   * <ul>
+   *   <li>In a block definition that comes sequentially before the block definition with the given
+   *       ID.
+   *   <li>In a block definition that has the same enumerator ID as the block definition with the
+   *       given ID.
+   * </ul>
+   */
+  public ImmutableList<QuestionDefinition> getAvailablePredicateQuestionDefinitions(long blockId)
+      throws ProgramBlockDefinitionNotFoundException {
+
+    Optional<Long> maybeEnumeratorId = getBlockDefinition(blockId).enumeratorId();
+    ImmutableList<BlockDefinition> blockDefinitionsWithSameEnumerator =
+        maybeEnumeratorId.isPresent()
+            ? getBlockDefinitionsForEnumerator(maybeEnumeratorId.get())
+            : getNonRepeatedBlockDefinitions();
+
+    ImmutableList.Builder<QuestionDefinition> resultBuilder = new ImmutableList.Builder<>();
+
+    for (BlockDefinition blockDefinition : blockDefinitionsWithSameEnumerator) {
```

Comment:
(nit): can you add to `#getBlockDefinitionsForEnumerator` docstring that the order is important and what the meaning of the order is?

Reply:
Oh yeah, let me do this.